### PR TITLE
Enabling mflog for metaflow

### PR DESCRIPTION
**Pull request by cloudw**
Enabling stream logging

### CODE REVIEW

File: metaflow/mflog/__init__.py [hunk 1]
```diff
@@ -37,6 +37,13 @@
             ' >> $MFLOG_STDOUT; echo $1; '\
      ' }' % TASK_LOG_SOURCE
 
+BASH_MFLOG_KFP =\
+    'mflog(){{ '\
+        'T=$(date -u -Ins|tr , .); '\
+        'echo \"[MFLOG|0|${{T:0:26}}Z|%s|$T]$1\"'\
+            ' >> $MFLOG_STDOUT; echo $1; '\
+     ' }}' % TASK_LOG_SOURCE
```
Comment (cloudw) [inline metaflow/mflog/__init__.py:45]: Double bracket and slightly different symbol escapes are needed to avoid bash command being unexpectedly formatted by python.

File: metaflow/plugins/kfp/kfp.py [hunk 2]
```diff
@@ -216,75 +225,69 @@ def _command(
         """
         Analogous to batch.py
         """
```
Comment (cloudw) [inline metaflow/plugins/kfp/kfp.py:230]: Reference: https://github.com/zillow/metaflow/blob/master/metaflow/plugins/aws/batch/batch.py#L48

File: metaflow/plugins/kfp/kfp_constants.py [hunk 3]
```diff
@@ -12,6 +12,6 @@
 TASK_ID_ENV_NAME = "TASK_ID_ENV_NAME"
 SPLIT_INDEX_ENV_NAME = "SPLIT_INDEX_ENV_NAME"
 INPUT_PATHS_ENV_NAME = "INPUT_PATHS_ENV_NAME"
-RETRY_COUNT = "RETRY_COUNT"
+RETRY_COUNT = "MF_ATTEMPT"
```
Comment (cloudw) [inline metaflow/plugins/kfp/kfp_constants.py:15]: Naming change by using `export_mflog_env_vars` over the `retry_count_python` bash command for setting env var.

File: metaflow/plugins/kfp/kfp.py [hunk 4]
```diff
@@ -69,6 +75,12 @@
     ScheduleDecorator,
 )
 
+LOGS_DIR = "/logs"
```
Comment (cloudw) [inline metaflow/plugins/kfp/kfp.py:78]: Should other location be used? This seems to be working for workflow example project. Maybe `metaflow/logs`?

Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:78]: yes, or and ideally in PVC shared folder if a PVC is used.
What is unix convention? maybe `/opt/metaflow_logs` ? and `/opt/metaflow_volume/metaflow_logs` if using a PVC, or just always  `/opt/metaflow_volume/metaflow_logs`

File: metaflow/plugins/kfp/kfp.py [hunk 5]
```diff
@@ -216,75 +228,71 @@ def _command(
...
+        # construct an entry point that
+        # 1) Clean attached volume if any
+        # 2) Initializes the mflog environment (mflog_expr)
+        # 3) Bootstraps a metaflow environment (init_expr)
+        # 4) Executes a task (step_expr)
+
+        # the `true` command is to make sure that the generated command
+        # plays well with docker containers which have entrypoint set as
+        # eval $@
+        cmd_str = f"{clean_volume} && mkdir -p /logs && {mflog_expr} && {init_expr} && {step_expr};"
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:283]: should use LOGS_DIR

File: metaflow/plugins/kfp/kfp_cli.py [hunk 6]
```diff
@@ -408,13 +396,16 @@ def make_flow(
...
+    package_url = None
+    if s3_code_package:
+        datastore = obj.datastore(
+                obj.flow.name,
+                mode="w",
+                metadata=obj.metadata,
+                event_logger=obj.event_logger,
+                monitor=obj.monitor,
+        )
+        package_url = datastore.save_data(obj.package.sha, TransformableObject(obj.package.blob))
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp_cli.py:408]: much cleaner, thank you!

**Review by talebzeghmi — CHANGES_REQUESTED** [review]
Great PR! This is nice and clean.

File: metaflow/metaflow_environment.py [hunk 7]
```diff
@@ -101,8 +101,9 @@ def get_boto3_copy_command(self, s3_path, local_path, command="download_file"):
             + "%s\"" % copy_command
         )
 
-    def get_package_commands(self, code_package_url, pip_install=True):
-        cmds = [BASH_MFLOG,
+    def get_package_commands(self, code_package_url, pip_install=True, kfp=False):
```
[the remark this answers is not in the document]
Comment (cloudw) [reply metaflow/metaflow_environment.py:104]: Will add doc string - just switching between two slightly different BASH_MFLOG command.

File: metaflow/mflog/__init__.py (hunk 1 above)
Comment (cloudw) [reply metaflow/mflog/__init__.py:45]: Example log line:
```
[MFLOG|0|2021-07-29T08:04:10.669355Z|task|8309088c-d273-4998-a7c5-a181009c91fc]Task is starting.
```
See the sample datastore for more examples 🙂.

File: metaflow/plugins/kfp/kfp.py (hunk 4 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:78]: Thanks for letting me know about the paths. Updating...

**Comment by cloudw** [comment]
Thank @talebzeghmi for pointing out KFP foreach support issue. Resolving and re-testing both foreach and retries...

File: metaflow/plugins/kfp/kfp.py [hunk 8]
```diff
@@ -216,75 +228,71 @@ def _command(
...
             'index = name.rfind("(");'
             'res = (0 if index == -1 else name[index + 1: -1]) if name.endswith(")") else 0;'
-            f'print("{RETRY_COUNT}=" + str(res))'
+            "print(str(res))"
         )
+        mflog_expr = export_mflog_env_vars(
+            flow_name=self.flow.name,
+            run_id="{run_id}",
+            step_name=step_name,
+            task_id=f"${TASK_ID_ENV_NAME}",
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:242]: oh this can be `task_id=TASK_ID_ENV_NAME,`

Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:242]: This might be rendering incorrectly... we don't know the task_id until runtime in the foreach case

File: metaflow/metaflow_environment.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (talebzeghmi) [reply metaflow/metaflow_environment.py:104]: The variable should convey intent, maybe `is_kfp_plugin`

File: metaflow/plugins/kfp/kfp_decorator.py [hunk 9]
```diff
@@ -106,26 +106,6 @@ def task_pre_step(
...
-                            ),
-                            "attempt": retry_count,
-                        }
-                    ),
-                    type="log_path",
-                    tags=[],
-                )
-            ]
-            # Register log related metadata for debugging.
-            metadata.register_metadata(run_id, step_name, task_id, datum)
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp_decorator.py:125]: where does the log metadata now get registered?

File: metaflow/plugins/kfp/tests/flows/foreach_linear_foreach.py [hunk 10]
```diff
@@ -34,15 +37,51 @@ def linear_4(self):
...
+            for data_input in self.x
+        ]
+        logs = []
+        print("Expect one of the following logs from each step: ", allowed_logs)
+        for task in logging_step.tasks():
+            log = task.stdout
+            print(f"=== Start: stdout from step index {task.index}===")
+            print(log)
+            print(f"=== End: stdout from step index {task.index}===")
+            assert any(log_candidate in log for log_candidate in allowed_logs)
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/tests/flows/foreach_linear_foreach.py:63]: This line doesn't validate that the log is correct, just that it is one of the `allowed_logs`.  
What if every task log was `Testing logging. Data: a`?  This could happen if the s3 paths are incorrect and overwrite each other.

File: metaflow/mflog/__init__.py (hunk 1 above)
Comment (talebzeghmi) [reply metaflow/mflog/__init__.py:45]: Please add as comments, thanks!

File: metaflow/plugins/kfp/kfp.py [hunk 11]
```diff
@@ -69,6 +73,12 @@
     ScheduleDecorator,
 )
 
+LOGS_DIR = "/opt/metaflow_volume/metaflow_logs"
+STDOUT_FILE = "mflog_stdout"
+STDERR_FILE = "mflog_stderr"
+STDOUT_PATH = os.path.join(LOGS_DIR, STDOUT_FILE)
+STDERR_PATH = os.path.join(LOGS_DIR, STDERR_FILE)
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:80]: These should be in `kfp_constants.py`

File: metaflow/plugins/kfp/kfp.py [hunk 12]
```diff
@@ -878,7 +900,7 @@ def kfp_pipeline_from_flow(
 
             def build_kfp_dag(
                 node: DAGNode,
-                passed_in_split_indexes: str = '""',
+                passed_in_split_indexes: str = "",
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:903]: why this change?

File: metaflow/plugins/kfp/kfp.py [hunk 13]
```diff
@@ -212,80 +222,91 @@ def _command(
...
+            f"&& {step_expr};"
         )
 
+        # after the task has finished, we save its exit code (fail/success)
+        # and persist the final logs. The whole entrypoint should exit
+        # with the exit code (c) of the task.
+        #
+        # Note that if step_expr OOMs, this tail expression is never executed.
+        # We lose the last logs in this scenario (although they are visible
+        # still through AWS CloudWatch console).
```
Comment (talebzeghmi) [inline metaflow/plugins/kfp/kfp.py:305]: Is this valid?

`
(although they are visible
        # still through AWS CloudWatch console)
`

Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:305]: > We lose the last logs in this scenario
What happens if the Flow crashes in the middle of execution, are logs still copied?  They were copied in our previous implementation.
This is a good test to have.

**Review by talebzeghmi — CHANGES_REQUESTED** [review]

File: metaflow/plugins/kfp/kfp.py (hunk 13 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:305]: Removing the AWS cloudwatch comment. Thanks

If a pod get ejected before flow finishes I don't think we still have logs in previous implementation.

File: metaflow/plugins/kfp/kfp.py (hunk 12 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:903]: The default value here is really saying no split index, which is much more intuitively an empty string. The quotation literals are only for bash code formatting and this formatting should be living in bash code. 

`passed_in_split_indexes` is also being used in mflog config code where I do want an empty string instead of a pair of quotation marks.

File: metaflow/metaflow_environment.py [hunk 14]
```diff
@@ -121,7 +124,7 @@ def get_package_commands(self, code_package_url, pip_install=True, kfp=False):
                     "after 6 tries. Exiting...\' && exit 1; "
                 "fi" % code_package_url,
                 "tar xf job.tar",
-                "mflog \'Task is starting.\'",
+                "mflog \'Task is starting.\'" if not is_kfp_plugin else "true",
```
[the remark this answers is not in the document]
Comment (cloudw) [reply metaflow/metaflow_environment.py:127]: For KFP there is another task starting announcement that happens after all bootstrap code actually finishes. Picking one of the two announcement the other more precisely announce the task start. The other announcement is here: https://github.com/zillow/metaflow/blob/967894dd0a0dd58628283e74fe24361981ccac22/metaflow/plugins/kfp/kfp.py#L271-L274

File: metaflow/metaflow_environment.py [hunk 15]
```diff
@@ -101,12 +101,15 @@ def get_boto3_copy_command(self, s3_path, local_path, command="download_file"):
...
+            self,
+            code_package_url,
+            is_kfp_plugin=False,
+    ):
+        mflog_bash_cmd = BASH_MFLOG if not is_kfp_plugin else BASH_MFLOG_KFP
         cmds = [mflog_bash_cmd,
                 "mflog \'Setting up task environment.\'",
-                "%s -m pip install click requests boto3 -qqq"
-                    % self._python() if pip_install else "true",  # true is Python pass for bash
+                "%s -m pip install click requests boto3 -qqq" % self._python(),
```
[the remark this answers is not in the document]
Comment (cloudw) [reply metaflow/metaflow_environment.py:112]: In current code `pip_install` is always set to `True` thus removing. Let's make it right in another PR then?

[the remark this answers is not in the document]
Comment (cloudw) [reply metaflow/metaflow_environment.py:112]: In what case is pip install not necessary? I think it's not necessary if a run is triggered from gitlab CI, or if user configure `pip_install` to be false.

If so `pip_install` should be passed from cli, and in gitlab CI we default it to `false` assuming project has a prebuilt image containing all necessary pacakges.

File: metaflow/plugins/kfp/kfp.py (hunk 13 above)
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:305]: And if the Flow fails within?

File: metaflow/metaflow_environment.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (talebzeghmi) [reply metaflow/metaflow_environment.py:112]: I didn't notice that it's always true:
https://github.com/zillow/metaflow/blob/736c72c45cf352da4ac8a850d90ed112630a63f4/metaflow/plugins/kfp/kfp.py#L224-L224

Good poing, let's leave it to another PR.

File: metaflow/plugins/kfp/kfp.py (hunk 13 above)
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py]: Manually tested RaiseErrorFlow and the exceptions are correctly captured: s3://serve-datalake-zillowgroup/zillow/workflow_sdk/metaflow_28d/dev/aip-example/RaiseErrorFlow/kfp-44b6944f-615e-49b0-9577-a485b500ea9a/error_step/kfp2/

Trying to automate this integration test now...

File: metaflow/metaflow_environment.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (cloudw) [reply metaflow/metaflow_environment.py:112]: Opened issue https://github.com/zillow/metaflow/issues/110

**Review by talebzeghmi — APPROVED** [review]
I'm glad you added the check on logs in failure test flow which cause that we aren't using the sidecar subprocces correctly!
